### PR TITLE
Implement search and history

### PR DESCRIPTION
**Pull request by anaerobeth**
Implements:
  - #search (holds the logic for resource, system and compartment searches)
  - #history (holds the logic for resource, type, and system history retrieval)

### CODE REVIEW

**Comment by bkaney** [comment]
I don't really understand the API here. I would assume `client.search(...)` should return the search based on heuristics. Not return a `level`?

File: lib/client.js [hunk 1]
```diff
@@ -489,6 +461,245 @@ class Client {
...
+   *
+   * // Using async
+   * let response = await fhirClient.history({ resourceType: 'Patient', id: '12345' });
+   * console.log(response);
+   *
+   * @param {Object} params - The request parameters.
+   * @param {string} params.resourceType - The resource type (e.g. "Patient", "Observation").
+   * @param {string} params.id - The FHIR id for the resource.
+   *
+   * @return {String} level - The history level
```
Comment (rserur) [inline lib/client.js]: History and search wouldn't return bundles?

**Comment by anaerobeth** [comment]
@bkaney I added `level` as a way to test what search variant was being called but Steve pointed out a way to move that logic to the test and preserve the search payload. We pushed two commits to fix the logic for `search` and `history` .

File: lib/client.js (hunk 1 above)
Comment (anaerobeth) [reply lib/client.js]: I pushed new commits to move the logic for checking search and history levels in the test, and restored the return value to `Bundle` in the relevant examples

File: lib/client.js [hunk 2]
```diff
@@ -489,6 +461,245 @@ class Client {
...
+   *   searchParams: { code: 'abc' }
+   * });
+   * console.log(response);
+   *
+   * @param {Object} params - The request parameters.
+   * @param {Object} params.searchParams - The search parameters.
+   *
+   * @return {Promise<Object>} FHIR resources in a FHIR Bundle structure.
+   */
+  compartmentSearch({ resourceType, compartment, searchParams } = {}) {
```
[the remark this answers is not in the document]
Comment (rserur) [reply lib/client.js]: ...but that question could always be answered by a later spike/discussion.

File: lib/client.js [hunk 3]
```diff
@@ -489,6 +461,229 @@ class Client {
...
+   * let response = await fhirClient.search({
+   *   resourceType: 'Patient',
+   *   compartment: { resourceType: 'Patient', id: 123 },
+   *   searchParams: { code: 'abc' },
+   * });
+   * console.log(response);
+   *
+   * @param {Object} params - The request parameters.
+   * @param {String} params.resourceType - The resource type (e.g. "Patient", "Observation").
+   * @param {Object} params.compartment - The search compartment.
```
Comment (bkaney) [inline lib/client.js]: Make sure to mark `[optional]` params

File: lib/client.js [hunk 4]
```diff
@@ -493,6 +464,234 @@ class Client {
   prevPage(results) {
     return this.pagination.prevPage(results);
   }
+
+  /**
+   * Search for a FHIR resource, within a compartment, or the entire system
```
Comment (Jammjammjamm) [inline lib/client.js]: Maybe this should say something like `with or without compartments` rather than `within a compartment`.

Comment (anaerobeth) [reply lib/client.js]: Updated in a new commit

File: lib/client.js [hunk 5]
```diff
@@ -493,6 +464,234 @@ class Client {
...
+   * // Using promises
+   * fhirClient.search({
+   *   resourceType: 'Observation',
+   *   compartment: { resourceType: 'Patient', id: 123 },
+   *   searchParams: { code: 'abc' },
+   * }).then((data) => { console.log(data); });
+   *
+   * // Using async
+   * let response = await fhirClient.search({
+   *   resourceType: 'Patient',
```
Comment (pdw207) [inline lib/client.js]: Observation

**Pull request merged**